runtime-sdk/modules/contracts/src/wasm.rs

```diff
@@ -0,0 +1,141 @@
+//! WASM runtime.
+use oasis_runtime_sdk::context::TxContext;
+
+use super::{
+    abi::{oasis::OasisV1, ExecutionContext, ExecutionOk, ABI},
+    types, Error, Parameters, MODULE_NAME,
+};
+
+/// Error emitted from within a contract.
+#[derive(Debug, thiserror::Error)]
+#[error("{message}")]
+pub struct ContractError {
+    pub module: String,
+    pub code: u32,
+    pub message: String,
+}
+
+impl ContractError {
+    /// Create a new error emitted within a contract.
+    pub fn new(code_id: types::CodeId, module: &str, code: u32, message: &str) -> Self {
+        Self {
+            module: if module.is_empty() {
+                format!("{}.{}", MODULE_NAME, code_id.as_u64())
+            } else {
+                format!("{}.{}.{}", MODULE_NAME, code_id.as_u64(), module)
+            },
+            code,
+            message: message.to_string(),
+        }
+    }
+}
+
+impl oasis_runtime_sdk::error::Error for ContractError {
+    fn module_name(&self) -> &str {
+        &self.module
+    }
+
+    fn code(&self) -> u32 {
+        self.code
+    }
+}
+
+/// Validate the passed contract code and compile it into an executable representation.
+pub(super) fn validate_and_compile<'ctx, C: TxContext>(
```

Comment:
the term `compile` feels a little disingenuous since there's really no lowering happening. right now it's more like `prepare_module` or something :)